src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -990,7 +990,23 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
                 }
             }
         } else if (ASRUtils::is_integer(*left_type) && ASRUtils::is_integer(*right_type)) {
-            dest_type = left_type;
+            int is_l64 = ASR::down_cast<ASR::Integer_t>(left_type)->m_kind == 8;
+            int is_r64 = ASR::down_cast<ASR::Integer_t>(right_type)->m_kind == 8;
+            if (is_l64 ^ is_r64) { // one is i64 and one is i32
+                if (is_l64) {
+                    dest_type = left_type;
+                    right = ASR::down_cast<ASR::expr_t>(ASR::make_ImplicitCast_t(
+                        al, right->base.loc, right, ASR::cast_kindType::IntegerToInteger, dest_type,
+                        value));
+                } else {
+                    dest_type = right_type;
+                    left = ASR::down_cast<ASR::expr_t>(ASR::make_ImplicitCast_t(
+                        al, right->base.loc, left, ASR::cast_kindType::IntegerToInteger, dest_type,
```

Comment:
I see. If you look at `ast_to_asr.cpp` in `lfortran` (see https://gitlab.com/lfortran/lfortran/-/blob/master/src/lfortran/semantics/asr_implicit_cast_rules.h and for usage example, see https://gitlab.com/lfortran/lfortran/-/blob/master/src/lfortran/semantics/ast_common_visitor.h#L145-150 ) implicit cast is completely abstracted out and is automated? Are you manually checking for inclusion of `ImplicitCast` in every node? If not, then probably you can abstract out implicit cast process as done above in LFortran so that code surface reduces and debugging efforts are optimized.

Reply:
Yeah, I see. Let me trying removing it.

Reply:
Removing this produces code generation error:
<details>
<summary> Error </summary>
<p>

```
Compiling...
+ ../src/bin/lpython test_math.py -o test_math
; ModuleID = 'LFortran'
source_filename = "LFortran"

@0 = private unnamed_addr constant [6 x i8] c"%lld\0A\00", align 1

define void @main0() {
.entry:
  %answer = alloca i64, align 8
  %x = alloca i32, align 4
  store i32 6, i32* %x, align 4
  %0 = call i64 @__module_modules_math_factorial(i32* %x)
  store i64 %0, i64* %answer, align 4
  %1 = load i64, i64* %answer, align 4
  call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @0, i32 0, i32 0), i64 %1)
  br label %return

return:                                           ; preds = %.entry
  ret void
}

define i64 @__module_modules_math_factorial(i32* %x) {
.entry:
  %_lpython_return_variable = alloca i64, align 8
  %result = alloca i64, align 8
  %0 = load i32, i32* %x, align 4
  %1 = icmp slt i32 %0, 0
  br i1 %1, label %then, label %else

then:                                             ; preds = %.entry
  store i32 0, i64* %_lpython_return_variable, align 4
  br label %return

unreachable_after_return:                         ; No predecessors!
  br label %ifcont

else:                                             ; preds = %.entry
  br label %ifcont

ifcont:                                           ; preds = %else, %unreachable_after_return
  store i32 1, i64* %result, align 4
  br label %loop.head

loop.head:                                        ; preds = %loop.body, %ifcont
  %2 = load i32, i32* %x, align 4
  %3 = icmp sgt i32 %2, 0
  br i1 %3, label %loop.body, label %loop.end

loop.body:                                        ; preds = %loop.head
  %4 = load i64, i64* %result, align 4
  %5 = load i32, i32* %x, align 4
  %6 = mul i64 %4, i32 %5
  store i64 %6, i64* %result, align 4
  %7 = load i32, i32* %x, align 4
  %8 = sub i32 %7, 1
  store i32 %8, i32* %x, align 4
  br label %loop.head

loop.end:                                         ; preds = %loop.head
  %9 = load i64, i64* %result, align 4
  store i64 %9, i64* %_lpython_return_variable, align 4
  br label %return

unreachable_after_return1:                        ; No predecessors!
  br label %return

return:                                           ; preds = %unreachable_after_return1, %loop.end, %then
  %10 = load i64, i64* %_lpython_return_variable, align 4
  ret i64 %10
}

declare void @_lfortran_printf(i8*, ...)

define i32 @main() {
.entry:
  ret i32 0
}
code generation error: asr_to_llvm: module failed verification. Error:
Stored value type does not match pointer operand type!
  store i32 0, i64* %_lpython_return_variable, align 4
 i64Stored value type does not match pointer operand type!
  store i32 1, i64* %result, align 4
 i64Both operands to a binary operator are not of the same type!
  %6 = mul i64 %4, i32 %5



Note: if any of the above error or warning messages are not clear or are lacking
context please report it to us (we consider that a bug that needs to be fixed).
Command '../src/bin/lpython test_math.py -o test_math' failed.

```
</p>
</details>

Reply:
N.B. - https://github.com/lcompilers/lpython/pull/103#discussion_r800077858

Reply:
Removing it won't work I think. You need to use `ImplicitCastRules` in your implementation. It's already used in `ast_to_asr.cpp` but NOT in `python_ast_to_asr.cpp`.